File: tools/packages/codeflow-utils-python/src/codeflow_utils/formatting/string.py
"""String formatting utilities."""

import re


def truncate_string(
    value: str,
    max_length: int,
    suffix: str = "...",
    preserve_words: bool = True,
) -> str:
    """
    Truncate string to maximum length.

    Args:
        value: String to truncate
        max_length: Maximum length (including suffix)
        suffix: Suffix to add if truncated
        preserve_words: Whether to preserve word boundaries

    Returns:
        Truncated string
    """
    if len(value) <= max_length:
        return value

    if preserve_words:
        # Try to truncate at word boundary
        truncated = value[:max_length - len(suffix)]
        last_space = truncated.rfind(" ")
        if last_space > max_length * 0.5:  # Only use word boundary if reasonable
            truncated = truncated[:last_space]
        return truncated + suffix

    return value[:max_length - len(suffix)] + suffix


def slugify(value: str, separator: str = "-") -> str:
    """
    Convert string to URL-friendly slug.

    Args:
        value: String to slugify
        separator: Word separator character

    Returns:
        Slugified string
    """
    # Convert to lowercase
    slug = value.lower()

    # Replace spaces and underscores with separator
    slug = re.sub(r"[\s_]+", separator, slug)

    # Remove all non-word characters except separator
    slug = re.sub(r"[^\w\-]+", "", slug)

    # Remove multiple separators
    slug = re.sub(rf"{re.escape(separator)}+", separator, slug)

    # Remove leading/trailing separators
    slug = slug.strip(separator)

    return slug


def camel_to_snake(value: str) -> str:
    """
    Convert camelCase to snake_case.

    Args:
        value: CamelCase string

    Returns:
        snake_case string
    """
    # Insert underscore before uppercase letters
    snake = re.sub(r"(?<!^)(?=[A-Z])", "_", value)
    return snake.lower()


def snake_to_camel(value: str, capitalize_first: bool = False) -> str:
    """
    Convert snake_case to camelCase.

    Args:
        value: snake_case string
        capitalize_first: Whether to capitalize first letter (PascalCase)

    Returns:
        camelCase or PascalCase string
    """
    components = value.split("_")

    if capitalize_first:
        return "".join(word.capitalize() for word in components)
    else:
        return components[0] + "".join(word.capitalize() for word in components[1:])

File: desktop/packages/utils/src/formatting/string.ts
/**
 * String formatting utilities.
 */

/**
 * Truncate string to maximum length.
 *
 * @param value - String to truncate
 * @param maxLength - Maximum length (including suffix)
 * @param suffix - Suffix to add if truncated
 * @param preserveWords - Whether to preserve word boundaries
 * @returns Truncated string
 */
export function truncateString(
  value: string,
  maxLength: number,
  suffix: string = "...",
  preserveWords: boolean = true
): string {
  if (value.length <= maxLength) {
    return value;
  }

  if (preserveWords) {
    const truncated = value.substring(0, maxLength - suffix.length);
    const lastSpace = truncated.lastIndexOf(" ");
    if (lastSpace > maxLength * 0.5) {
      return truncated.substring(0, lastSpace) + suffix;
    }
    return truncated + suffix;
  }

  return value.substring(0, maxLength - suffix.length) + suffix;
}

/**
 * Convert string to URL-friendly slug.
 *
 * @param value - String to slugify
 * @param separator - Word separator character
 * @returns Slugified string
 */
export function slugify(value: string, separator: string = "-"): string {
  return value
    .toLowerCase()
    .trim()
    .replace(/[\s_]+/g, separator)
    .replace(/[^\w\-]+/g, "")
    .replace(new RegExp(`${separator}+`, "g"), separator)
    .replace(new RegExp(`^${separator}|${separator}$`, "g"), "");
}

/**
 * Convert camelCase to snake_case.
 *
 * @param value - CamelCase string
 * @returns snake_case string
 */
export function camelToSnake(value: string): string {
  return value
    .replace(/([A-Z])/g, "_$1")
    .toLowerCase()
    .replace(/^_/, "");
}

/**
 * Convert snake_case to camelCase.
 *
 * @param value - snake_case string
 * @param capitalizeFirst - Whether to capitalize first letter (PascalCase)
 * @returns camelCase or PascalCase string
 */
export function snakeToCamel(
  value: string,
  capitalizeFirst: boolean = false
): string {
  const components = value.split("_");
  const first = capitalizeFirst
    ? components[0].charAt(0).toUpperCase() + components[0].slice(1)
    : components[0];
  const rest = components
    .slice(1)
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1));
  return first + rest.join("");
}
